### ISSUE TITLE
Can you multi select with Ag?

### ISSUE THREAD

**Issue by tobymao** [junegunn/fzf.vim#40]
I would really like to be able to multi select after doing :Ag in vim. Is there an easy way to get this to work?

**Comment by junegunn** [comment]
It's already possible:
- `tab` - select and move cursor down
- `shift-tab` - select and move cursor up (for some reason, neovim doesn't support the key)
- `ctrl-a` - select all
- `ctrl-d` - deselect all
- `enter` / `ctrl-x` / `ctrl-v` / `ctrl-t` - open selected entries (current window / horizontal splits / vertical splits / tabs)

Selected entries are added to quickfix list.

**Comment by tobymao** [comment]
Sorry, I wasn't clear enough. When I select multiple things with tab and then press ctrl-t, I expected it to open multiple tabs like the :FZF -m command does.

**Comment by junegunn** [comment]
That is certainly pretty trivial to implement.

```diff
diff --git a/autoload/fzf/vim.vim b/autoload/fzf/vim.vim
index 957cb8e..bb9c700 100644
--- a/autoload/fzf/vim.vim
+++ b/autoload/fzf/vim.vim
@@ -387,15 +387,17 @@ function! s:ag_handler(lines)
   let cmd = get(get(g:, 'fzf_action', s:default_action), a:lines[0], 'e')
   let list = map(a:lines[1:], 's:ag_to_qf(v:val)')
 
-  let first = list[0]
-  execute cmd s:escape(first.filename)
-  execute first.lnum
-  execute 'normal!' first.col.'|zz'
+  for entry in list
+    execute cmd s:escape(entry.filename)
+    execute entry.lnum
+    execute 'normal!' entry.col.'|zz'
+  endfor
 
   if len(list) > 1
     call setqflist(list)
     copen
     wincmd p
+    clast
   endif
 endfunction
 ```

However, the thing is there are times when you want to select tens of, or hundreds of matches and build quickfix list out of them, especially with `CTRL-A` key binding. And opening hundreds of tabs or splits is of course not ideal. I think we can add bang-version of the command, i.e. `Ag!` that has this alternative behavior.

**Comment by tobymao** [comment]
What about like Ag -m or some other option?

**Comment by junegunn** [comment]
`-m` I think is a bit ambiguous. I'll think about the options, meanwhile you can fork the repo and apply the patch above.

**Comment by tobymao** [comment]
Ok, thank you very much.

**Comment by nhooyr** [comment]
Isn't this fixed already?

**Comment by junegunn** [comment]
Nope. It's about adding an option to make `:Ag` open multiple tabs or splits instead of populating quickfix list.

**Comment by john-kurkowski** [comment]
This issue applies to any call to `fzf#vim#grep`. `:Ag` uses it. Custom grep programs besides `ag` also use it.

To clarify the issue, when you multi-select from `fzf#vim#grep`, it seems only the _1st_ result honors the `Ctrl-{T,V,X}` shortcut. _All_ selected results, including the 1st, make it into the quickfix list however. There's no way out of populating the quickfix.

I think the solution is #185, and to stop making `:Ag` multi-select a special case.

**Comment by felipesere** [comment]
This is biting me at the moment. I don't care too much ab out the quicklist to be honest.
Could we add a global setting to use the quicklist (or not) `g:fzf_use_quickfix`? I rarely open more than 5 files and at a time from the search and going through quickfix is annoying.

Alternatively a cutoff could be added to use the quickfix if for example more than 10 files are selected...

**Comment by dylan-chong** [comment]
Shouldn't AG respect the fzf_action global variable? I just realised that this command also does not respect fzf_action which appears to be related

```
command! -bang -nargs=* Rg
      \ call fzf#vim#grep(
      \   "rg --column --line-number --no-heading --color=always --smart-case -g '!.git' --hidden ".shellescape(<q-args>), 1,
      \   fzf#vim#with_preview('right:35%'),
      \   <bang>0)

```

**Comment by nadim** [comment]
Current workaround is to load everything into quickfix and then I have a separate function that opens all files from that list.

**Comment by DieracDelta** [comment]
Is it possible to select an entry from the list and open it in a new tab while still keeping the fzf window open? My usecase here is I'd like to open a file, look at it for a couple of seconds, then close it and return to my search. The search takes a fairly long time so I'd really only like to run it once.

**Comment by mshtawythug** [comment]
> My current workaround is to load everything into quickfix and then I have a separate function that opens all files from that list.
> 
> Source: https://pastebin.com/J9RwciFQ
> 
> ```
> " quickfixopenall.vim
> "Author:
> "   Tim Dahlin
> "
> "Description:
> "   Opens all the files in the quickfix list for editing.
> "
> "Usage:
> "   1. Perform a vimgrep search
> "       :vimgrep /def/ *.rb
> "   2. Issue QuickFixOpenAll command
> "       :QuickFixOpenAll
> 
> function!   QuickFixOpenAll()
>     if empty(getqflist())
>         return
>     endif
>     let s:prev_val = ""
>     for d in getqflist()
>         let s:curr_val = bufname(d.bufnr)
>         if (s:curr_val != s:prev_val)
>             exec "edit " . s:curr_val
>         endif
>         let s:prev_val = s:curr_val
>     endfor
> endfunction
> 
> command! QuickFixOpenAll         call QuickFixOpenAll()
> ```

I change just one line to I can open in vertical split
      exec "vert sb " . s:curr_val

**Comment by mshtawythug** [comment]
> My current workaround is to load everything into quickfix and then I have a separate function that opens all files from that list.
> 
> Source: https://pastebin.com/J9RwciFQ
> 
> ```
> " quickfixopenall.vim
> "Author:
> "   Tim Dahlin
> "
> "Description:
> "   Opens all the files in the quickfix list for editing.
> "
> "Usage:
> "   1. Perform a vimgrep search
> "       :vimgrep /def/ *.rb
> "   2. Issue QuickFixOpenAll command
> "       :QuickFixOpenAll
> 
> function!   QuickFixOpenAll()
>     if empty(getqflist())
>         return
>     endif
>     let s:prev_val = ""
>     for d in getqflist()
>         let s:curr_val = bufname(d.bufnr)
>         if (s:curr_val != s:prev_val)
>             exec "edit " . s:curr_val
>         endif
>         let s:prev_val = s:curr_val
>     endfor
> endfunction
> 
> command! QuickFixOpenAll         call QuickFixOpenAll()
> ```

@nadim I liked your solution, I modified it to open in a new tab
```
function! QuickFixOpenAll()
  if empty(getqflist())
    return
  endif
  let modifier = "tabe"
  let s:prev_val = ""
  for d in getqflist()
    let s:curr_val = bufname(d.bufnr)
    if (s:curr_val != s:prev_val)
      echom modifier." " . s:curr_val
      exec modifier." " . s:curr_val
      let modifier = 'vert sb'
    endif
    let s:prev_val = s:curr_val
  endfor
endfunction

command! QuickFixOpenAll call QuickFixOpenAll()
nnoremap <leader>oa :QuickFixOpenAll<cr>
```
